File: libs/native-raw-tcp-printer/android/src/main/java/com/mysmallbusiness/nativerawtcpprinter/NativeRawTcpPrinterModule.kt
package com.mysmallbusiness.nativerawtcpprinter

import android.graphics.Bitmap
import android.graphics.Canvas
import android.view.View
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import java.net.InetSocketAddress
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap

private data class Raster(val width: Int, val height: Int, val rgba: ByteArray)
private data class Options(val viewTag: Int, val host: String, val port: Int, val width: Int, val copies: Int, val captureScale: Int, val timeoutMs: Int)

private object PrinterQueues {
  private val queues = ConcurrentHashMap<String, Mutex>()
  suspend fun <T> run(key: String, block: suspend () -> T): T = queues.getOrPut(key) { Mutex() }.withLock { block() }
}

class NativeRawTcpPrinterModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("NativeRawTcpPrinter")
    AsyncFunction("print") Coroutine { input: Map<String, Any?> ->
      val started = System.nanoTime()
      try {
        val options = parseOptions(input)
        val captureStarted = System.nanoTime()
        val bitmap = withContext(Dispatchers.Main) { capture(options.viewTag, options.captureScale) }
        val capturedAt = System.nanoTime()
        val resized = resize(bitmapToRgba(bitmap), options.width)
        val resizedAt = System.nanoTime()
        val bytes = escPos(resized)
        val rasterAt = System.nanoTime()
        var sendMs = 0L
        val sendStarted = System.nanoTime()
        PrinterQueues.run(options.host.trim().lowercase() + ":" + options.port) {
          withContext(Dispatchers.IO) { send(options, bytes) }
        }
        sendMs = elapsedMs(sendStarted)
        mapOf(
          "ok" to true, "captureMs" to elapsedMs(captureStarted, capturedAt), "resizeMs" to elapsedMs(capturedAt, resizedAt),
          "rasterMs" to elapsedMs(resizedAt, rasterAt), "sendMs" to sendMs, "totalMs" to elapsedMs(started),
          "width" to resized.width, "height" to resized.height, "byteLength" to bytes.size, "fnv1a32" to fnv1a32(bytes), "sent" to true
        )
      } catch (error: PrinterFailure) {
        mapOf("ok" to false, "error" to mapOf("code" to error.code, "message" to error.message.orEmpty(), "phase" to error.phase), "totalMs" to elapsedMs(started))
      } catch (error: Throwable) {
        mapOf("ok" to false, "error" to mapOf("code" to "RASTER_FAILED", "message" to (error.message ?: "Native printer failure"), "phase" to "native"), "totalMs" to elapsedMs(started))
      }
    }
  }

  private fun parseOptions(input: Map<String, Any?>): Options {
    fun number(name: String): Int = (input[name] as? Number)?.toInt() ?: throw PrinterFailure("INVALID_OPTIONS", "options", "$name is required")
    val host = input["host"] as? String ?: throw PrinterFailure("INVALID_OPTIONS", "options", "host is required")
    val result = Options(number("viewTag"), host, number("port"), number("width"), number("copies"), number("captureScale"), number("timeoutMs"))
    if (result.host.isBlank() || result.port !in 1..65535 || result.width < 8 || result.copies !in 1..10 || result.captureScale !in 1..2) throw PrinterFailure("INVALID_OPTIONS", "options", "Invalid printer options")
    return result
  }

  private fun capture(tag: Int, captureScale: Int): Bitmap {
    val view = appContext.findView<View>(tag) ?: throw PrinterFailure("VIEW_NOT_FOUND", "capture", "Receipt view was not found")
    if (view.width <= 0 || view.height <= 0) throw PrinterFailure("CAPTURE_FAILED", "capture", "Receipt view has no size")
    return Bitmap.createBitmap(view.width * captureScale, view.height * captureScale, Bitmap.Config.ARGB_8888).also { Canvas(it).apply { scale(captureScale.toFloat(), captureScale.toFloat()); view.draw(this) } }
  }
}

private class PrinterFailure(val code: String, val phase: String, message: String) : Exception(message)
private fun elapsedMs(start: Long, end: Long = System.nanoTime()) = (end - start) / 1_000_000
private fun bitmapToRgba(bitmap: Bitmap): Raster {
  val pixels = IntArray(bitmap.width * bitmap.height); bitmap.getPixels(pixels, 0, bitmap.width, 0, 0, bitmap.width, bitmap.height)
  val rgba = ByteArray(pixels.size * 4)
  pixels.forEachIndexed { i, pixel -> val o = i * 4; rgba[o] = (pixel shr 16).toByte(); rgba[o + 1] = (pixel shr 8).toByte(); rgba[o + 2] = pixel.toByte(); rgba[o + 3] = (pixel ushr 24).toByte() }
  return Raster(bitmap.width, bitmap.height, rgba)
}
private fun resize(source: Raster, requestedWidth: Int): Raster {
  val maxWidth = requestedWidth.coerceAtLeast(8) / 8 * 8
  if (source.width <= maxWidth && source.width % 8 == 0) return source
  val scale = minOf(1.0, maxWidth.toDouble() / source.width); val width = maxOf(8, (source.width * scale).toInt()) / 8 * 8; val height = maxOf(1, (source.height * scale).toInt())
  val output = ByteArray(width * height * 4)
  for (y in 0 until height) for (x in 0 until width) { val from = ((minOf(source.height - 1, (y / scale).toInt()) * source.width + minOf(source.width - 1, (x / scale).toInt())) * 4); val to = (y * width + x) * 4; for (c in 0..3) output[to + c] = source.rgba[from + c] }
  return Raster(width, height, output)
}
private fun escPos(raster: Raster): ByteArray {
  val output = ArrayList<Byte>(); fun put(value: Int) { output.add(value.toByte()) }; fun putAll(values: ByteArray) { values.forEach { output.add(it) } }
  putAll(byteArrayOf(0x1b, 0x40, 0x1b, 0x61, 0x01)); putAll(byteArrayOf(0x1b, 0x33, 24))
  for (y in 0 until raster.height step 24) { put(0x1b); put(0x2a); put(33); put(raster.width and 0xff); put(raster.width shr 8)
    for (x in 0 until raster.width) for (stripe in 0..2) { var value = 0; for (bit in 0..7) { val py = y + stripe * 8 + bit; if (py < raster.height) { val i = (py * raster.width + x) * 4; val alpha = raster.rgba[i + 3].toInt() and 0xff; val lum = (raster.rgba[i].toInt() and 0xff) * .299 + (raster.rgba[i + 1].toInt() and 0xff) * .587 + (raster.rgba[i + 2].toInt() and 0xff) * .114; if (alpha > 0 && lum < 180) value = value or (1 shl (7 - bit)) } }; put(value) }
    put(0x0a)
  }
  putAll(byteArrayOf(0x1b, 0x32, 0x0a, 0x0a, 0x0a, 0x1d, 0x56, 0x00)); return output.toByteArray()
}
private fun send(options: Options, bytes: ByteArray) { try { repeat(options.copies) { Socket().use { socket -> socket.connect(InetSocketAddress(options.host, options.port), options.timeoutMs); socket.soTimeout = options.timeoutMs; socket.getOutputStream().use { it.write(bytes); it.flush() } } } } catch (error: java.net.SocketTimeoutException) { throw PrinterFailure("TIMEOUT", "send", error.message ?: "Printer timed out") } catch (error: Throwable) { throw PrinterFailure("SEND_FAILED", "send", error.message ?: "Unable to send to printer") } }
private fun fnv1a32(bytes: ByteArray): String { var hash = 0x811c9dc5.toInt(); bytes.forEach { hash = hash xor (it.toInt() and 0xff); hash *= 0x01000193 }; return (hash.toLong() and 0xffffffffL).toString(16).padStart(8, '0') }

File: libs/native-raw-tcp-printer/ios/NativeRawTcpPrinterModule.swift
import ExpoModulesCore
import Network
import UIKit
import CryptoKit

private struct PrinterOptions {
  let viewTag: Int; let host: String; let port: UInt16; let width: Int; let copies: Int; let captureScale: CGFloat; let timeoutMs: Int
}
private struct Raster { let width: Int; let height: Int; let rgba: [UInt8] }
private struct NativeFailure: Error { let code: String; let phase: String; let message: String }

private actor PrinterQueues {
  static let shared = PrinterQueues()
  private var tails: [String: Task<Void, Never>] = [:]
  func run<T: Sendable>(key: String, _ work: @escaping @Sendable () async throws -> T) async throws -> T {
    let previous = tails[key]
    let task = Task { if let previous { _ = await previous.result } }
    tails[key] = task
    _ = await task.result
    defer { if tails[key] == task { tails.removeValue(forKey: key) } }
    return try await work()
  }
}

public final class NativeRawTcpPrinterModule: Module {
  public func definition() -> ModuleDefinition {
    Name("NativeRawTcpPrinter")
    AsyncFunction("print") { (input: [String: Any]) async -> [String: Any] in
      let started = DispatchTime.now().uptimeNanoseconds
      do {
        let options = try Self.parse(input)
        let captureStart = DispatchTime.now().uptimeNanoseconds
        let image = try await MainActor.run { try self.capture(tag: options.viewTag, scale: options.captureScale) }
        let captureEnd = DispatchTime.now().uptimeNanoseconds
        let raster = try Self.resize(try Self.rgba(image), maxWidth: options.width)
        let resizeEnd = DispatchTime.now().uptimeNanoseconds
        let bytes = Self.escPos(raster)
        let rasterEnd = DispatchTime.now().uptimeNanoseconds
        var sendMs: UInt64 = 0
        let sendStart = DispatchTime.now().uptimeNanoseconds
        let target = options.host.lowercased() + ":" + String(options.port)
        try await PrinterQueues.shared.run(key: target) { try await Self.send(bytes, options: options) }
        sendMs = Self.elapsed(sendStart)
        return ["ok": true, "captureMs": Self.elapsed(captureStart, captureEnd), "resizeMs": Self.elapsed(captureEnd, resizeEnd), "rasterMs": Self.elapsed(resizeEnd, rasterEnd), "sendMs": sendMs, "totalMs": Self.elapsed(started), "width": raster.width, "height": raster.height, "byteLength": bytes.count, "fnv1a32": Self.fnv1a32(bytes), "sent": true]
      } catch let failure as NativeFailure {
        return ["ok": false, "error": ["code": failure.code, "phase": failure.phase, "message": failure.message], "totalMs": Self.elapsed(started)]
      } catch {
        return ["ok": false, "error": ["code": "RASTER_FAILED", "phase": "native", "message": error.localizedDescription], "totalMs": Self.elapsed(started)]
      }
    }
  }

  private func capture(tag: Int, scale: CGFloat) throws -> UIImage {
    guard let view: UIView = appContext?.findView(withTag: tag, ofType: UIView.self) else { throw NativeFailure(code: "VIEW_NOT_FOUND", phase: "capture", message: "Receipt view was not found") }
    guard view.bounds.width > 0 && view.bounds.height > 0 else { throw NativeFailure(code: "CAPTURE_FAILED", phase: "capture", message: "Receipt view has no size") }
    let renderer = UIGraphicsImageRenderer(bounds: view.bounds, format: { let format = UIGraphicsImageRendererFormat(); format.scale = scale; return format }())
    return renderer.image { _ in view.drawHierarchy(in: view.bounds, afterScreenUpdates: true) }
  }

  private static func parse(_ value: [String: Any]) throws -> PrinterOptions {
    guard let viewTag = value["viewTag"] as? Int, let host = value["host"] as? String, let port = value["port"] as? Int, let width = value["width"] as? Int, let copies = value["copies"] as? Int, let timeoutMs = value["timeoutMs"] as? Int else { throw NativeFailure(code: "INVALID_OPTIONS", phase: "options", message: "Missing printer options") }
    guard let captureScale = value["captureScale"] as? Double, !host.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, (1...65535).contains(port), width >= 8, (1...10).contains(copies), timeoutMs > 0, (1...2).contains(captureScale) else { throw NativeFailure(code: "INVALID_OPTIONS", phase: "options", message: "Invalid printer options") }
    return PrinterOptions(viewTag: viewTag, host: host, port: UInt16(port), width: width, copies: copies, captureScale: CGFloat(captureScale), timeoutMs: timeoutMs)
  }

  private static func rgba(_ image: UIImage) throws -> Raster {
    guard let cgImage = image.cgImage else { throw NativeFailure(code: "CAPTURE_FAILED", phase: "capture", message: "Receipt image is unavailable") }
    let width = cgImage.width; let height = cgImage.height; var pixels = [UInt8](repeating: 0, count: width * height * 4)
    guard let context = CGContext(data: &pixels, width: width, height: height, bitsPerComponent: 8, bytesPerRow: width * 4, space: CGColorSpaceCreateDeviceRGB(), bitmapInfo: CGImageAlphaInfo.premultipliedLast.rawValue | CGBitmapInfo.byteOrder32Big.rawValue) else { throw NativeFailure(code: "CAPTURE_FAILED", phase: "capture", message: "Unable to read receipt pixels") }
    context.draw(cgImage, in: CGRect(x: 0, y: 0, width: width, height: height))
    return Raster(width: width, height: height, rgba: pixels)
  }

  private static func resize(_ source: Raster, maxWidth requested: Int) -> Raster {
    let maxWidth = max(8, requested) / 8 * 8
    if source.width <= maxWidth && source.width % 8 == 0 { return source }
    let scale = min(1, Double(maxWidth) / Double(source.width)); let width = max(8, Int(Double(source.width) * scale)) / 8 * 8; let height = max(1, Int(Double(source.height) * scale)); var output = [UInt8](repeating: 0, count: width * height * 4)
    for y in 0..<height { for x in 0..<width { let sx = min(source.width - 1, Int(Double(x) / scale)); let sy = min(source.height - 1, Int(Double(y) / scale)); let from = (sy * source.width + sx) * 4; let to = (y * width + x) * 4; output[to..<to+4] = source.rgba[from..<from+4] } }
    return Raster(width: width, height: height, rgba: output)
  }

  private static func escPos(_ raster: Raster) -> [UInt8] {
    var bytes: [UInt8] = [0x1b, 0x40, 0x1b, 0x61, 0x01, 0x1b, 0x33, 24]
    for y in stride(from: 0, to: raster.height, by: 24) { bytes += [0x1b, 0x2a, 33, UInt8(raster.width & 0xff), UInt8(raster.width >> 8)]
      for x in 0..<raster.width { for stripe in 0..<3 { var value: UInt8 = 0; for bit in 0..<8 { let py = y + stripe * 8 + bit; if py < raster.height { let i = (py * raster.width + x) * 4; let luminance = Double(raster.rgba[i]) * 0.299 + Double(raster.rgba[i+1]) * 0.587 + Double(raster.rgba[i+2]) * 0.114; if raster.rgba[i+3] > 0 && luminance < 180 { value |= UInt8(1 << (7 - bit)) } } }; bytes.append(value) } }; bytes.append(0x0a) }
    bytes += [0x1b, 0x32, 0x0a, 0x0a, 0x0a, 0x1d, 0x56, 0x00]; return bytes
  }

  private static func send(_ bytes: [UInt8], options: PrinterOptions) async throws {
    for _ in 0..<options.copies { try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
      let connection = NWConnection(host: NWEndpoint.Host(options.host), port: NWEndpoint.Port(rawValue: options.port)!, using: .tcp)
      var settled = false; func finish(_ result: Result<Void, Error>) { guard !settled else { return }; settled = true; connection.cancel(); continuation.resume(with: result) }
      connection.stateUpdateHandler = { state in if case .failed(let error) = state { finish(.failure(NativeFailure(code: "CONNECTION_FAILED", phase: "send", message: error.localizedDescription))) } }
      connection.start(queue: .global()); connection.send(content: Data(bytes), completion: .contentProcessed { error in if let error { finish(.failure(NativeFailure(code: "SEND_FAILED", phase: "send", message: error.localizedDescription))) } else { finish(.success(())) } })
      DispatchQueue.global().asyncAfter(deadline: .now() + .milliseconds(options.timeoutMs)) { finish(.failure(NativeFailure(code: "TIMEOUT", phase: "send", message: "Printer timed out"))) }
    } }
  }

  private static func elapsed(_ start: UInt64, _ end: UInt64 = DispatchTime.now().uptimeNanoseconds) -> UInt64 { (end - start) / 1_000_000 }
  private static func fnv1a32(_ bytes: [UInt8]) -> String { var hash: UInt32 = 0x811c9dc5; for byte in bytes { hash ^= UInt32(byte); hash = hash &* 0x01000193 }; return String(format: "%08x", hash) }
}
